src/io/grail.jl

```diff
@@ -0,0 +1,262 @@
+                    gm_consumers[consumer_index] = gm_consumer
+                    consumer_count += 1
+                else
+                    #=
+                    gm_producer = Dict{String,Any}(
+                        "index" => producer_count,
+                        "qg_junc" => junction_id,
+                        "qgmax" => 0.0,
+                        "qgmin" => 0.0,
+                        "qgfirm" => -withdrawal
+                    )
+                    =#
+
+                    gm_producer = Dict{String,Any}(
+                        "index" => producer_count,
+                        "qg_junc" => junction_id,
+                        "qgmax" => -withdrawal,
+                        "qgmin" => 0.0,
+                        "qgfirm" => 0.0
+                    )
+
+                    producer_index = "$(gm_producer["index"])"
+                    assert(!haskey(gm_producers, producer_index))
+                    gm_producers[producer_index] = gm_producer
+                    producer_count += 1
+                end
+            end
+        end
+
+        if node["isslack"] != 0 && slack_producers
+            warn("adding producer at junction $(junction_id) to model slack capacity")
+
+            gm_producer = Dict{String,Any}(
+                "index" => producer_count,
+                "qg_junc" => junction_id,
+                "qgmax" => node["qmax"],
+                "qgmin" => node["qmin"],
+                "qgfirm" => 0.0
+            )
+
+            producer_index = "$(gm_producer["index"])"
+            assert(!haskey(gm_producers, producer_index))
+            gm_producers[producer_index] = gm_producer
+            producer_count += 1
+        end
+
+    end
+
+    #println(length(gm_junctions))
+    #println(length(gm_producers))
+    #println(length(gm_consumers))
+
+    gm_connections = Dict{String,Any}()
+    for (i, edge) in g_edges
+        assert(edge["fr_node"] == node_id_to_junction_id[edge["f_id"]])
+        assert(edge["to_node"] == node_id_to_junction_id[edge["t_id"]])
+
+        # assume diameter units in inches, convert to meters
+        # assume length units is in degrees, convert to meters
+        a = 343.0 # speed of sound constant
+        #a = 1.0
+
+        length = max(edge["length"], 0.1) # to be robust to zero values
+
+        c = 96.074830e-15            # Gas relative constant
+        L = length*54.0*1.6  # length of the pipe [km]
+        D = edge["diameter"]*25.4    # interior diameter of the pipe [mm]
+        T = 281.15                   # gas temperature [K]
+        epsilon = 0.05               # absolute rugosity of pipe [mm]
+        delta = 0.6106               # density of the gas relative to air [-]
+        z = 0.8                      # gas compressibility factor [-]
+        B = 3.6*D/epsilon
+        lambda = 1/((2*log10(B))^2)
+        resistance = c*(D^5/(lambda*z*T*L*delta));
+
+        resistance = max(resistance, 0.01) # to have numerical robustness
```

Comment:
@rb004f please take note of the two max functions I am using here; is case I am doing something very crazy.